### PR TITLE
add option of saving pid file

### PR DESCRIPTION
**Pull request by hbl0307106015**
Signed-off-by: Tymon  <banglang.huang@foxmail.com>

   I think wifidog-gateway should have a option to save pid file, so I add this function.

### CODE REVIEW

**Comment by hbl0307106015** [comment]
The build is successful in my openwrt-trunk, could you please show me more hints ?

**Comment by acv** [comment]
So there's two things here.

Technically, the code does not fail to build. But the non-SSL integration build are set to convert all warnings to errors. The SSL build don't run with that compiler flag so they succeed.

The error itself is that getpid() returns pid_t which is a signed long on linux. Yet the printf() specifies an unsigned long. %zu might be the right string to use here, I believe that current OpenWrt use a C99 compliant toolchain.

**Comment by hbl0307106015** [comment]
thank you for suggestions, I will commit another patch for it later.

**Comment by hbl0307106015** [comment]
I 'd already modified the commit just now, please check it.

**Comment by mhaas** [comment]
Hey, travis (or rather, the C compiler) does not like some of your code. Can you check that and fix the warning?

**Comment by hbl0307106015** [comment]
I can not see the Travis CI build details, could you show me more information so that I can correct my code ?

**Comment by hbl0307106015** [comment]
Of course I really want to know where is the warning occur.

**Comment by mhaas** [comment]
You can see the build log here: https://travis-ci.org/wifidog/wifidog-gateway/jobs/72659411

**Comment by hbl0307106015** [comment]
Ok, I got it, all build error has been fixed.

File: src/util.c [hunk 1]
```diff
@@ -389,3 +389,21 @@ rand16(void)
...
+/*
+ * Save pid of this wifidog in pid file
+ * @param 'pf' as string, it is the pid file absolutely path
+ */
+void
+save_pid_file(const char *pf)
+{
+    if (pf) {
+        FILE *f = fopen(pf, "w");
+        if (f) {
```
Comment (mhaas) [inline src/util.c]: Can you provide a meaningful error message to the user if the fopen call fails?

File: src/util.c [hunk 2]
```diff
@@ -389,3 +389,21 @@ rand16(void)
...
+            fprintf(f, "%d\n", getpid());
+            fclose(f);
```
Comment (mhaas) [inline src/util.c]: We probably need to check the return value for fclose as well.

**Comment by mhaas** [comment]
Much appreciated. I added some comments to your patch - we probably need to check the return values. Otherwise coverity will complain because we're not programming defensively.

**Comment by hbl0307106015** [comment]
well, thanks mhaas' suggestions, I add some meaningful message to user and check the return value of fclose as well.

File: src/util.c [hunk 3]
```diff
@@ -389,3 +389,27 @@ rand16(void)
...
+        if (f) {
+            fprintf(f, "%d\n", getpid());
+
+            int ret = fclose(f);
+            if (ret == EOF) /* check the return value of fclose */
+                debug(LOG_ERR, "fclose() on file %s was failed (%s)", pf, strerror(errno));
+        }
+
+        /* fopen return NULL, open file failed */
+        debug(LOG_ERR, "fopen() on flie %s was failed (%s)", pf, strerror(errno));
```
Comment (mhaas) [inline src/util.c]: I believe there is a small control flow error here: the call do `debug` call will always fire regardless of the value of `f`.

Comment (hbl0307106015) [inline src/util.c]: I am sorry to make a stupid mistake.

**Comment by hbl0307106015** [comment]
Anything needed to be mofificated ?

**Comment by mhaas** [comment]
No, it's perfect now. Thanks for the PR and implementing the changes!

**Pull request merged**